packages/react/dismissable-layer/src/DismissableLayer.tsx

```diff
@@ -1,241 +1,167 @@
 
-// We need to compute the total count of layers AND a running count of all layers
-// in order to find which layer is the deepest one.
-// This is use to only dismiss the deepest layer when using the escape key
-// because we bind the key listener to document so cannot take advantage of event.stopPropagation()
-const [TotalLayerCountProvider, useTotalLayerCount] = createTotalLayerCount();
-const [RunningLayerCountProvider, usePreviousRunningLayerCount] = createRunningLayerCount();
-
-// We need to compute the total count of layers which set `disableOutsidePointerEvents` to `true` AND
-// a running count of all the layers which set `disableOutsidePointerEvents` to `true` in order to determine
-// which layers should be dismissed when interacting outside.
-// (ie. all layers that do not have a child layer which sets `disableOutsidePointerEvents` to `true`)
-const [
-  TotalLayerCountWithDisabledOutsidePointerEventsProvider,
-  useTotalLayerCountWithDisabledOutsidePointerEvents,
-] = createTotalLayerCount('TotalLayerCountWithDisabledOutsidePointerEventsProvider');
-const [
-  RunningLayerCountWithDisabledOutsidePointerEventsProvider,
-  usePreviousRunningLayerCountWithDisabledOutsidePointerEvents,
-] = createRunningLayerCount('RunningLayerCountWithDisabledOutsidePointerEventsProvider');
-
 /* -------------------------------------------------------------------------------------------------
  * DismissableLayer
  * -----------------------------------------------------------------------------------------------*/
 
 const DISMISSABLE_LAYER_NAME = 'DismissableLayer';
+const CONTEXT_UPDATE = 'dismissableLayer.update';
+const POINTER_DOWN_OUTSIDE = 'dismissableLayer.pointerDownOutside';
+const FOCUS_OUTSIDE = 'dismissableLayer.focusOutside';
 
-type DismissableLayerElement = DismissableLayerImplElement;
-interface DismissableLayerProps extends DismissableLayerImplProps {}
-
-const DismissableLayer = React.forwardRef<DismissableLayerElement, DismissableLayerProps>(
-  (props, forwardedRef) => {
-    const runningLayerCount = usePreviousRunningLayerCount();
-    const isRootLayer = runningLayerCount === 0;
-    const layer = <DismissableLayerImpl {...props} ref={forwardedRef} />;
-
-    // if it's the root layer, we wrap it with our necessary root providers
-    // (effectively we wrap the whole tree of nested layers)
-    return isRootLayer ? (
-      <TotalLayerCountProvider>
-        <TotalLayerCountWithDisabledOutsidePointerEventsProvider>
-          {layer}
-        </TotalLayerCountWithDisabledOutsidePointerEventsProvider>
-      </TotalLayerCountProvider>
-    ) : (
-      layer
-    );
-  }
-);
-
-DismissableLayer.displayName = DISMISSABLE_LAYER_NAME;
-
-/* -----------------------------------------------------------------------------------------------*/
+const DismissableLayerContext = React.createContext({
+  layers: new Set<DismissableLayerElement>(),
+  modals: new Set<DismissableLayerElement>(),
```

Comment:
It really bothers me that we use "modality" throughout to refer to the layers that disable outside pointer events because at this stage it's not yet about modality.

At the same time, I can see why you went for that because it's so much smaller than `layersWithDisabledOutsidePointerEvents` which is more the language I was using previously…

I'm torn!

Reply:
Maybe we could use a different term like `blockingLayers` and refer to "blocking" throughout?

Reply:
i find `blocking` more confusing tbh, blocking what? 😬

the term modality encapsulates that things cannot be clicked outside of them so seems clearer there but i get that it could imply more than that too.

i'll see if i can come up with something else.

Reply:
Yeah it's one of the criteria of modality, but not the entirety of it I guess that's what is bothering me.